### PR TITLE
Experimental TPU implementation of DistributedDataParallel

### PR DESCRIPTION
**Pull request by will-cromar**
- Adds experimental implementation of DDP that works on TPU v3 and gives moderately better performance on TPU v4 (1805 ex/sec/replica vs 1710 ex/sec/replica in my experiments).
- Add C++ methods to get PJRT process index and process count (similar to JAX)
- Initialize `torch.distributed` in `pjrt.spawn`, since the master address and process ranks are not predictable on TPU.
- Tweak learning rate and `atol` in DDP tests. The LR was so small that the parameter comparison always passed, even when I had errors in my code.
- Don't broadcast buffers in ResNet50 example to match native XLA distributed implementation.
- Test PJRT with both upstream and experimental DDP.

Tested with PJRT and XRT:

```
XRT_TPU_CONFIG="localservice;0;localhost:51011" TPU_NUM_DEVICES=4 MASTER_ADDR=localhost MASTER_PORT=12355 python pytorch/xla/test/test_ddp.py
PJRT_DEVICE=TPU python pytorch/xla/test/pjrt/test_ddp.py
```

### CODE REVIEW

File: test/distributed_util.py [hunk 1]
```diff
@@ -93,11 +90,13 @@ def train_step(model, inputs, labels, optimizer, loss_fn):
...
-def ddp_correctness(init_file: str,
-                    *,
+def ddp_correctness(ddp: type = torch.nn.parallel.DistributedDataParallel,
                     use_large_net: bool = False,
                     debug: bool = False):
-  rank, world_size = init_xla_backend(init_file)
+  if not dist.is_initialized():
+    rank, world_size = init_xla_backend()
+  else:
+    rank, world_size = xm.get_ordinal(), xm.xrt_world_size()
```
Comment (JackCaoG) [inline test/distributed_util.py:99]: one customer asked if they can use `torch.distributed` to get the ordinal and world size, do you know if that will work if we init the xla backend?

File: torch_xla/experimental/pjrt.py [hunk 2]
```diff
@@ -179,9 +192,25 @@ def _thread_fn(device: torch.device):
...
+
+    if os.getenv('PJRT_INIT_TORCH_DISTRIBUTED', '0') == '1':
+      if device_type() == 'TPU':
+        # HACK: need to call with each device since it relies on an XLA collective
+        master_ip = next(executor.map(_discover_tpu_master_worker_ip, devices))
+        init_method = f'tcp://{master_ip}:{master_port}'
+      else:
+        init_method = None
+
+      init_pjrt_process_group(init_method=init_method)
```
Comment (JackCaoG) [inline torch_xla/experimental/pjrt.py:212]: hmm, so user don't need to init the process group if they are using our wrapper?

File: test/distributed_util.py (hunk 1 above)
Comment (JackCaoG) [reply test/distributed_util.py:99]: if I understand correctly
1. if user use the upstream ddp, they need to manually `init_process_group`
2. if user use our custom wrapper, they don't.

If this a fair understanding?

Comment (will-cromar) [reply test/distributed_util.py:99]: Ordinal and world size will be correct in `torch.distributed` on TPU v4 if the process group is initialized with `init_pjrt_process_group`, but not on v2/v3, because we'll have two ordinals per rank on v2/v3.

I added the option to initialize the torch.distributed group through `xmp.spawn` for three reasons, in order of importance:

1. The master IP is not predictable in a TPU slice.
2. If on TPU v2/v3, only one thread can initialize the `torch.distributed` group. It's easier to do it here than tell the user to manage that themselves.
3. It may be convenient for the PJRT process ID to match each process' rank so we have one fewer index to keep track of.

So, with either DDP implementation, the user can initialize the main process group themselves if they solve the three problems above.

The default is to not initialize the process group since we found that TCP process group initialization is flaky on large pod slices.

File: torch_xla/experimental/pjrt.py (hunk 2 above)
Comment (will-cromar) [reply torch_xla/experimental/pjrt.py:212]: See above.

**Review by JackCaoG — APPROVED** [review]
Can we update README and add corresponding tests on xlml? I think we don't have either native ddp nor this new custom ddp e2e test running right now.

File: test/test_train_mp_imagenet.py [hunk 3]
```diff
@@ -193,7 +200,7 @@ def train_imagenet():
   device = xm.xla_device()
   model = get_model_property('model_fn')().to(device)
   if FLAGS.ddp:
-    model = DDP(model, gradient_as_bucket_view=True)
+    model = DDP(model, gradient_as_bucket_view=True, broadcast_buffers=False)
```
Comment (alanwaketan) [inline test/test_train_mp_imagenet.py:203]: Why do we disable this option?

File: torch_xla/experimental/pjrt.py [hunk 4]
```diff
@@ -311,3 +340,61 @@ def rendezvous(tag: str, payload: bytes,
...
+
+  @staticmethod
+  def _reduce_grad(grad: torch.Tensor) -> torch.Tensor:
+    """Average gradients across replicas."""
+    return xm.all_reduce(xm.REDUCE_SUM, grad, scale=1. / global_device_count())
+
+  def __init__(self,
+               module: nn.Module,
+               *,
+               broadcast_buffers: bool = True,
```
Comment (alanwaketan) [inline torch_xla/experimental/pjrt.py:381]: I guess we don't need to implement this logic?

**Review by alanwaketan — APPROVED** [review]
LGTM. Thanks, Will. This looks awesome.

File: torch_xla/experimental/pjrt.py (hunk 4 above)
Comment (will-cromar) [reply torch_xla/experimental/pjrt.py:381]: The `broadcast_buffers` is implemented in `forward`

**Comment by will-cromar** [comment]
Filed bugs to update documentation and add tests.

File: torch_xla/experimental/pjrt.py (hunk 4 above)
Comment (alanwaketan) [reply torch_xla/experimental/pjrt.py:381]: I guess my confusion is like if the previous optimizer_step approach doesn't demand this, what's the benefits of providing such option here? And then it doesn't seem like you test this option.
